test/e2e/basic/ping-pong.go

```diff
@@ -2,160 +2,168 @@ package e2e
 		opts := metav1.ListOptions{}
-		pods, err := cs.Pods("ping-pong").List(context.TODO(), opts)
+		pods, err := cs.Pods(pingPongName).List(context.Background(), opts)
 		gomega.Expect(err).NotTo(gomega.HaveOccurred())
 
 		for _, pod := range pods.Items {
-			var log string
-			//run command in pod
-			ginkgo.By("Ensuring that ping-pong is working")
-			if pod.Status.Phase == v1.PodRunning {
-				log, err = GetPodLogs(pod.Namespace, pod.Name, "", false)
-				gomega.Expect(err).NotTo(gomega.HaveOccurred())
-				if !strings.Contains(log, "Ping") || !strings.Contains(log, "Pong") {
-					warn.OnError(errors.New("Did not see Ping or either Pong, waiting"))
-				}
-
-				if strings.Contains(log, "Ping") && strings.Contains(log, "Pong") {
-					ginkgo.By("Found Ping, Pong in logs, done")
-					return
-				}
+			log, err := getPodLogs(clientSet, pod.Namespace, pod.Name)
+			if err != nil {
+				return err
+			}
+			if !strings.Contains(log, "Ping") || !strings.Contains(log, "Pong") {
+				continue
+			}
+
+			if strings.Contains(log, "Ping") && strings.Contains(log, "Pong") {
+				ginkgo.By("Found Ping, Pong in logs, done")
+				return nil
 			}
 		}
 	}
-
 }
 
-func specialResourceDelete(cs *framework.ClientSet, cl client.Client, path string) {
-	SRName := "ping-pong"
-	ginkgo.By("deleting " + SRName)
-	sr, err := ioutil.ReadFile(path)
+func pingPongDelete(cs *framework.ClientSet, cl client.Client) error {
+	sr, err := ioutil.ReadFile(pingPongChartPath)
 	if err != nil {
-		panic(err)
+		return err
 	}
-	framework.DeleteFromYAMLWithCR(sr, cl)
+	return framework.DeleteFromYAMLWithCR(sr, cl)
+}
 
-	ginkgo.By(fmt.Sprintf("Confirming %s is deleted", SRName))
-	err = wait.PollImmediate(pollInterval, waitDuration, func() (bool, error) {
+func waitPingPongDeleted(cl client.Client) error {
+	err := wait.PollImmediate(pollInterval, waitDuration, func() (bool, error) {
 		specialresources := &srov1beta1.SpecialResourceList{}
-		err = cl.List(context.TODO(), specialresources, []client.ListOption{}...)
+		err := cl.List(context.Background(), specialresources, []client.ListOption{}...)
 		if err != nil {
-			return false, fmt.Errorf("couldn't get SpecialResourceList: %v", err)
+			return false, err
 		}
-
 		for _, n := range specialresources.Items {
-			if n.ObjectMeta.Name == SRName {
+			if n.ObjectMeta.Name == pingPongName {
 				return false, nil
 			}
 		}
-
 		return true, nil
 
 	})
-	if err != nil {
-		explain = err.Error()
-	}
-	gomega.Expect(err).NotTo(gomega.HaveOccurred(), explain)
+	return err
 }
 
-func specialResourceCreate(cs *framework.ClientSet, cl client.Client, path string) {
-
-	ginkgo.By("creating ping-pong")
-	sr, err := ioutil.ReadFile(path)
+func waitPingPongReady(clientSet *kubernetes.Clientset) error {
+	watch, err := clientSet.CoreV1().Pods(pingPongName).Watch(context.Background(), metav1.ListOptions{})
 	if err != nil {
-		panic(err)
+		return err
 	}
-	framework.CreateFromYAML(sr, cl)
+	quitChan := make(chan struct{})
```

Comment:
How about using a [`sync.WaitGroup`](https://pkg.go.dev/sync#WaitGroup) here instead?
```suggestion
	var wg sync.WaitGroup
	wg.Add(2)
```
Then instead of `numPodsRunning++`, `wg.Done()`
Finally, `wg.Wait()` instead of waiting on the channel.

Reply:
Swapped it with wait groups, but I still need to count to exit the go routine.

Reply:
I think you can do `wg.Done()` two times instead of counting in a a variable?

Reply:
If I use only one go routine I need to know when the condition is met to exit, and that means counting. I can add `wg.Add(2)` and then call `wg.Done()` twice, but I still to place a return somewhere, and that can only happen once you have seen two pods running. All this while listening to events, since there is a watch in that namespace. What do you suggest as exit condition?

Reply:
Hmmm, I missed that aspect - you indeed need to stop the running goroutine.  
One way I can think of would be to have a cancellable context created at the very beginning of the function. The context would be passed to `clientSet.CoreV1().Pods(pingPongName).Watch()` and the cancellation function would be cancelled after `wg.Wait()` returns.
In the goroutine, a `select` statement would be used to wait on both `watch.ResultChan()` and `ctx.Done()`.  
This idea, although a bit more convoluted, brings the benefit of automatically cancelling `clientSet.CoreV1().Pods(pingPongName).Watch()`, which otherwise [seems to be done by the server every 5 minutes or so](https://github.com/kubernetes/client-go/issues/623).

Reply:
Neat! It is done now.